apps/freeablo/main.cpp

```diff
@@ -31,29 +31,96 @@ bool paused = false;
 bool noclip = false;
 int changeLevel = 0;
 
-int quit_key;
-int noclip_key;
-int changelvldwn_key;
-int changelvlup_key;
+int hotkey [5];
+
+int quit_key [5];
+int noclip_key [5];
+int changelvldwn_key [5];
+int changelvlup_key [5];
+
+bpt::ptree hotkeypt;
 
 void keyPress(Input::Key key)
 {
-    if (key == quit_key)
+    switch(key)
+    {
+        case Input::KEY_RSHIFT: hotkey[1] = 1; return;
+        case Input::KEY_LSHIFT: hotkey[1] = 1; return;
+        case Input::KEY_RCTRL: hotkey[2] = 1; return;
+        case Input::KEY_LCTRL: hotkey[2] = 1; return;
+        case Input::KEY_RALT: hotkey[3] = 1; return;
+        case Input::KEY_LALT: hotkey[3] = 1; return;
+        case Input::KEY_RSUPER: hotkey[4] = 1; return;
+        case Input::KEY_LSUPER: hotkey[4] = 1; return;
+        default:
+            {
+                hotkey[0] = key;
+                break;
+            }
+    }
+    
+    int check = 0;
```

Comment:
The stuff you are doing here with check is confusing. It would be better to create a hotkey class which holds a Key member, and a member for each modifer, and then just use comparison.